src/state_machine/serial.rs

```diff
@@ -68,3 +76,65 @@ where
         Ok(Rc::new(T::decode(target)?))
     }
 }
+
+impl<T: Encode> Encode for Vec<T> {
+    fn encode(&self, target: &mut impl std::io::Write) {
+        target
+            .write_all(&(self.len() as u64).to_le_bytes())
+            .expect("Serialization should not fail");
+        for item in self.iter() {
+            item.encode(target)
+        }
+    }
+}
+
+impl<T: Decode<Error = E>, E> Decode for Vec<T>
+where
+    DeserializationError: From<E>,
```

Comment:
I think we shouldn't assume an `Error` type for any type T.

What about:
1. Moving the length-checking logic to a separate method and checking it before decoding.
2. Defining a distinct type, `PrefixedVec', then we could check the length in a constructor, so the error is emitted as soon as we get the vec not when we try to deserialize it.

Reply:
Hmm, it would be nice to allow decoding of `Vec<T>` directly, but let me think about the mechanism. 

I guess one thing we could do would be to swap DeserializationError for a more generic error type like `CodecError`, which implements From: std::error::Error.

Reply:
Canceling this in favor of #18